### PR TITLE
Integrate Multimodal tools to `agentskills`.

### PR DESCRIPTION
**Pull request by Shimada666**
@xingyaoww @mczhuge
This PR is associated with #1941 #1914. 
The purpose is to integrate some readers into agentskill, providing the agent with additional capabilities. 
To prevent agentskill from becoming too bloated, I have removed some readers, retaining only those that I believe are necessary. Readers related to parsing binary files will be prioritized for retention. The current list of retained readers is subject to discussion. Below are detailed descriptions of the situation.

## Retained and implemented
PDFReader, DOCXReader,  PPTXReader, AudioReader, ImageReader, VideoReader

## Retained but subject to discussion
LaTexReader: Retained, but I'm not sure if it is needed.

## Removed
TXTReader, JSONReader, JSONLReader, XMLReader, YAMLReader, HTMLReader: Removed, replaced with open_file.
ExcelReader, XLSXReader: Removed, replaced with pandas
ZIPReader:  Removed, agent can just unzip in a bash shell
PythonReader: Removed, agent can run code

### CODE REVIEW

File: pyproject.toml
Comment (xingyaoww) [inline pyproject.toml:35]: Can you move these to the test dependencies? Since we will be using these inside the docker sandbox AND when running unit tests.

Also, can you add pip install of these libraries to the Dockerfile of our sandbox so that it is by default installed?

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (xingyaoww) [inline opendevin/runtime/plugins/agent_skills/agentskills.py:29]: Make sure you add the `pip install` commands for these libraries to the `runtime/plugins/agent_skills/setup.sh`

File: poetry.lock
Comment (yufansong) [inline poetry.lock:168]: Some package version seems do not need to change. You can try `poetry lock --no-update` to update the `poetry.lock` file

**Review by yufansong — COMMENTED** [review]
Maybe you can revert some change in the `poetry.lock` file.

File: poetry.lock
Comment (Shimada666) [reply poetry.lock:168]: Thank you! I'll give it a try.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [inline opendevin/runtime/plugins/agent_skills/agentskills.py:37]: This will not work out of the box because those environmental variables are not available in the sandbox.

We need to rethink about the design here (including how to mock those API calls in the tests), but I think for now you could put those environmental variables into the sandbox in `__init__` function of `sandbox.py`

**Review by li-boxuan — COMMENTED** [review]
Could you please provide an example/demo of CodeAct using any of these new skills?

**Comment by li-boxuan** [comment]
We are not using `runtime` to interact with the file system... ideally we should not call any os APIs, but I think that's fine for now... `runtime` APIs are not being used by any code yet it seems? C.C. @rbren

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:37]: I think these tools are primarily needed for GAIA (https://github.com/OpenDevin/OpenDevin/pull/1911) to perform well as proposed in https://github.com/OpenDevin/OpenDevin/pull/1914

In the default case, we should just disable these tools for the agent, and only enable them for evals now. In the case for GAIA evaluation, we can just set `SANDBOX_ OPENAI_API_KEY`.

I discussed with @rbren earlier, I think it will a more fundamental solution if we can connect these `agentskills` (executed inside sandbox) with `Event Stream` on our backend (e.g., every time a audio-to-text function is called by the agent inside sandbox, it push an `AudioToTextAction` to the `EventStream` on the backend - maybe by exposing a REST API from backend; The python function execution inside sandbox will be blocked until an `AudioToTextObservation` is processed by the backend and returned to the function in the sandbox).

**Comment by xingyaoww** [comment]
@li-boxuan As discuss above about the potential next step architecture (REST API that allows sandbox to push actions to the EventStream), we can rely on pure Python for file operation now, and then gradually replacing them with our defined file operations (e.g., send `FileEditAction` to the Event stream when `edit_file` function is called, and simply wait for the result).

This gives us the best of both world: the flexibility of code execution (you can compose multiple tools together really easily) AND the structured outputs required for great user interface (user need to know what the agent is doing).

**Comment by xingyaoww** [comment]
> Could you please provide an example/demo of CodeAct using any of these new skills?

I view this is more like a dependency for GAIA (https://github.com/OpenDevin/OpenDevin/pull/1911) ;)

File: containers/sandbox/Dockerfile
Comment (xingyaoww) [reply containers/sandbox/Dockerfile:35]: I actually think if there's a way for us to build two sandboxes, one can be a minimal one that is used for most agents that do not require these; another one can be tagged ":agent_skills" version, which includes this dependency.
I think installing a lot of these packages for `agentskills` on every sandbox initialization just take too much time - would be nice to have a specific version of container that already has these dependencies installed

Comment (li-boxuan) [reply containers/sandbox/Dockerfile:35]: Yeah two sandboxes sounds like a plan

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:37]: > In the case for GAIA evaluation, we can just set SANDBOX_OPENAI_API_KEY.

Okay yeah that makes sense for evaluation purposes

Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:37]: Yep i'm actually hoping we can have a way to optionally switch on/off these tools - so that they won't get included into the prompt for normal tasks

File: containers/sandbox/Dockerfile
Comment (li-boxuan) [reply containers/sandbox/Dockerfile:35]:
```suggestion
RUN pip3 install jupyterlab notebook jupyter_kernel_gateway flake8
# TODO: those dependencies are needed for agentskills, we should pack them in a new sandbox image
RUN pip3 install python-docx PyPDF2 openpyxl python-pptx pylatexenc requests pandas openai opencv-python
```

Comment (xingyaoww) [reply containers/sandbox/Dockerfile:35]: @Shimada666 could you see if we could achieve this easily? Thanks!!!

Comment (Shimada666) [reply containers/sandbox/Dockerfile:35]: If these are not installed in the Dockerfile in advance, installing the dependencies for agentskill will be relatively slow, probably 20+ seconds, affect the user experience. Of course, the image size has also increased significantly, 545M -> 1G...

In fact, I plan to merge agentskill with Jupyter later on, leaving only agentskill plugins. At that time, installing these dependencies will also be essential.

Alternatively, we can temporarily disable large packages, such as pandas and opencv, as discussed in the architecture below. These audio and video readers are only temporary solutions here; ultimately, interaction will rely on the server side. The sandbox will simply send a command to the REST server, and the specific invocation will be executed by the REST server.

Comment (xingyaoww) [reply containers/sandbox/Dockerfile:35]: In the worst case, for simplicity, we can just not install these dependencies in the sandbox at all, since we have PIP cache mechanism.

But I'm kinda still leaning toward getting all those in at the moment and gradually refractor out these utilities to our action library (e.g., PDF reader), since the CodeActAgent at the main branch currently default to use `AgentSkills` so pre-installed dependencies could save some hassle.

Comment (li-boxuan) [reply containers/sandbox/Dockerfile:35]: Yeah I can live with the status quo as long as we leave a comment for future reference

**Comment by li-boxuan** [comment]
I just tested that I could use CodeActAgent to read a pdf locally and summarise its content. Neat!

image

**Review by li-boxuan — APPROVED** [review]

File: containers/sandbox/Dockerfile
Comment (Shimada666) [reply containers/sandbox/Dockerfile]: Test Result

In the original state, installing dependencies takes 2 seconds
Excluding the opencv package, installing dependencies takes 8 seconds. If included in the image, the image size increases by 50M.
Including all the packages, installing dependencies takes 10 seconds. If included in the image, the image size increases by 168M.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (xingyaoww) [inline opendevin/runtime/plugins/agent_skills/agentskills.py:44]: Actually, can we add a check here: If `OPENAI_API_KEY` is NOT set, we remove the functions that requires `client` (`parse_audio`, `parse_image`, `parse_audio`) from the documentation?

File: pyproject.toml
Comment (xingyaoww) [inline pyproject.toml:42]: Can you remove these dependencies, now you have moved them to the test section?

**Review by xingyaoww — COMMENTED** [review]
Mostly LGTM! Just a few small issues that need to be fixed before we merge it

File: pyproject.toml
Comment (yufansong) [reply pyproject.toml]: removed.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (yufansong) [reply opendevin/runtime/plugins/agent_skills/agentskills.py:44]: done

**Comment by yufansong** [comment]
Good, I try it locally. It can successfully open a pdf and get content correctly.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (li-boxuan) [inline opendevin/runtime/plugins/agent_skills/agentskills.py:613]: Aha, I believe this has led to the integration test failure, but that makes sense... let me push updates to fix it

**Review by xingyaoww — APPROVED** [review]
LGTM and tested to make sure it worked for PDF, pptx, and latex!

PDF:
image

LaTeX:
image

PPTX:
image

**Comment by mczhuge** [comment]
SUPER COOL. And my tears gone.

**Comment by Shimada666** [comment]
@li-boxuan @xingyaoww @yufansong @mczhuge 
Today I was not at home, and when I came back, I saw that the PR had been refined and merged... Thank you all so much!!! Thank you for your help!!! 🥺🥺

File: opendevin/runtime/plugins/agent_skills/setup.sh
Comment (SmartManoj) [inline opendevin/runtime/plugins/agent_skills/setup.sh:13]: Do you think this line is necessary when it has already been added to the docker file?

Comment (li-boxuan) [reply opendevin/runtime/plugins/agent_skills/setup.sh:13]: Maybe not? This could be easily verified.

Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/setup.sh:13]: Yes - i think we should better keep this. We will want to support when the user brings their own docker image: https://github.com/OpenDevin/OpenDevin/issues/1387

When that happens, we should check whether the dependencies are installed - if not, we should install them.

Also check this PR: https://github.com/OpenDevin/OpenDevin/pull/2101

Comment (xingyaoww) [reply opendevin/runtime/plugins/agent_skills/setup.sh:13]: This pip install will just be a `no-op` when it is already installed, so it doesn't hurt i think.

Comment (rbren) [reply opendevin/runtime/plugins/agent_skills/setup.sh:13]: Agreed--we want to try and support non-default docker images, so making sure the plugins take care of all their own dependencies is good.

Comment (SmartManoj) [reply opendevin/runtime/plugins/agent_skills/setup.sh:13]: `RuntimeError: Failed to initialize plugin jupyter with exit code 1 and output: The folder you are executing pip from can no longer be found.`
Getting weird errors in CI with persistent sandbox.
https://github.com/SmartManoj/Kevin/actions/runs/9330781998/job/25684728084

**Pull request merged**
